test/unit/pipeline/FetchUnitTest.cc

```diff
@@ -442,5 +462,288 @@ TEST_F(PipelineFetchUnitTest, idleLoopBufferDueToNotTakenBoundary) {
+        .WillByDefault(Return(insnMaxSizeBytes));
+    ON_CALL(isa, getMinInstructionSize())
+        .WillByDefault(Return(insnMinSizeBytes));
+    ON_CALL(memory, getCompletedReads()).WillByDefault(Return(completedReads));
+
+    // Buffer will contain valid max and min sized instruction, predecode
+    // returns max bytes read on first tick
+    MacroOp mOp = {uopPtr};
+    ON_CALL(isa, predecode(_, insnMaxSizeBytes + insnMinSizeBytes,
+                           0x10 - (insnMaxSizeBytes + insnMinSizeBytes), _))
+        .WillByDefault(DoAll(SetArgReferee<3>(mOp), Return(insnMaxSizeBytes)));
+
+    // Fetch the data, only last max + min bytes from block. Should allow
+    // continuation to predecode
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    EXPECT_CALL(memory, getCompletedReads()).Times(1);
+    EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+    EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+    uint64_t setPC = blockSize - (insnMaxSizeBytes + insnMinSizeBytes);
+    // Fetch a minimum and maximum sized instruction, buffered bytes = 0
+    fetchUnit.updatePC(setPC);
+    // Tick and predecode max bytes
+    fetchUnit.tick();
+
+    // Ensure max bytes consumed
+    EXPECT_EQ(fetchUnit.bufferedBytes_, insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.pc_, blockSize - insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp);
+
+    // Expect that memory is requested even though there is data in the buffer
+    // as bufferedBytes < maxInstructionSize
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    EXPECT_CALL(memory,
+                requestRead(Field(&memory::MemoryAccessTarget::address, 16), _))
+        .Times(1);
+    fetchUnit.requestFromPC();
+
+    EXPECT_EQ(fetchUnit.bufferedBytes_, insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.pc_, blockSize - insnMinSizeBytes);
+
+    // Memory doesn't complete reads in next cycle but buffered bytes should be
+    // predecoded
+    MacroOp mOp2 = {uopPtr2};
+    ON_CALL(memory, getCompletedReads())
+        .WillByDefault(Return(span<memory::MemoryReadResult>{nullptr, 0}));
+    ON_CALL(isa, predecode(_, insnMinSizeBytes, 0x10 - insnMinSizeBytes, _))
+        .WillByDefault(DoAll(SetArgReferee<3>(mOp2), Return(insnMinSizeBytes)));
+
+    // Path through fetch as follows:
+    // More data required as bufferedBytes_ < maxInsnSize so getCompletedReads
+    // Doesn't complete so buffer doesn't get added to
+    // Buffer still has some valid data so predecode should be called
+
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    EXPECT_CALL(memory, getCompletedReads()).Times(1);
+    EXPECT_CALL(isa, getMinInstructionSize()).Times(2);
+    EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+    // Tick
+    fetchUnit.tick();
+
+    // Ensure min bytes are consumed
+    EXPECT_EQ(fetchUnit.bufferedBytes_, 0);
+    EXPECT_EQ(fetchUnit.pc_, 16);
+    EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp2);
+  }
```

Comment:
I think rather than address 16, we should base the next fetch address off of the block size